### PR TITLE
Don't use KUBE-MARK-DROP for LoadBalancerSourceRanges

### PR DESCRIPTION
**Pull request by danwinship**
#### What type of PR is this?
/kind cleanup
/kind feature

#### What this PR does / why we need it:
As part of [KEP-3178](https://github.com/kubernetes/enhancements/issues/3178) this removes the use of `KUBE-MARK-DROP` in the handling of the service LoadBalancerSourceRanges feature.

Old behavior:
- If `LoadBalancerSourceRanges` is enabled, then traffic to the LB IP goes to a `KUBE-FW-*` chain that matches each allowed source range and jumps to the `KUBE-EXT-*` chain on a match.
- If none of the source ranges match, the last rule in the `KUBE-FW-*` chain calls `KUBE-MARK-DROP` to mark the packet to be dropped.
- Later, in the `filter` table, the `KUBE-FIREWALL` chain sees the mark and (regardless of what else has happened to the packet in the meantime), calls `-j DROP` on the packet
- (Note that the short-circuit to `ClusterIP` doesn't happen until the `KUBE-EXT-*` chain, so pod-to-LB-IP packets that don't match any `LoadBalancerSourceRanges` ranges will get dropped, not short-circuited.)

New behavior:
- [SAME] If `LoadBalancerSourceRanges` is enabled, then traffic to the LB IP goes to a `KUBE-FW-*` chain that matches each allowed source range and jumps to the `KUBE-EXT-*` chain on a match.
- If none of the source ranges match, the `KUBE-FW-*` chain (implicitly) returns. No other rules in the `nat` table should end up matching the packet, so it will arrive in the `filter` table un-rewritten.
- In the `filter` table, a rule in ~`KUBE-EXTERNAL-SERVICES`~ `KUBE-PROXY-FIREWALL` will see the packet still addressed to the LB IP and call `-j DROP` on it.
  - (`KUBE-EXTERNAL-SERVICES` doesn't work because it only runs from `INPUT` and `FORWARD`, whereas we need to intercept `OUTPUT` too in this case to catch disallowed pod-to-LB-IP traffic. So we create a new firewall chain (with the same name as the one added for ipvs KEP-3178.)
- [SAME] (Note that the short-circuit to `ClusterIP` doesn't happen until the `KUBE-EXT-*` chain, so pod-to-LB-IP packets that don't match any `LoadBalancerSourceRanges` ranges will get dropped, not short-circuited.)

#### Which issue(s) this PR fixes:
none

#### Special notes for your reviewer:
In the past, the e2e test that is supposed to make sure this feature works didn't actually work (#85572) (in the sense that the test would pass even if we didn't explicitly drop the packet), and I think that is still true with this PR; if we failed to drop the packet, it would just get blackholed rather than actually being incorrectly delivered to a service endpoint. It is possible that there is not (and never has been) any situation where this DROP rule is actually _needed_...

At any rate, it should be possible to manually test this by doing:
1. create a LoadBalancer service with LoadBalancerSourceRanges that allow connections from outside the cluster but not from node/pod IPs
2. create a test pod
3. on the test pod's node, `kill -STOP` kube-proxy to prevent it from deleting and recreating rules while we're trying to monitor their counts
4. on the test pod's node, `iptables-save -c -t filter| grep <LB-IP>` to see the current counter on the drop rule
5. from the test pod, try to connect to the LB IP
6. on the test pod's node, `iptables-save -c -t filter| grep <LB-IP>` to confirm that the counter on the rule went up

This actually doesn't work in OpenShift, because of changes we made to the way that GCP VIPs work to make apiserver failover work better during upgrades, which apparently have side effects that affect local-to-LB-IP connections that don't get matched by any kube-proxy rules. (The packet still gets dropped, it just doesn't get dropped by the right rule.) So I need to fix that, but even then, OpenShift's behavior won't be identical to "normal" GCP, so I still need someone to test this on "normal" GCP. (**EDIT**: actually, probably testing on bare metal with MetalLB is beter anyway, so I need to test that. **EDIT 2**: and now tested)

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3178
```

### CODE REVIEW

**Comment by danwinship** [comment]
> > I still need someone to test this on "normal" GCP.
> 
> /test pull-kubernetes-e2e-gci-gce-ingress

as mentioned above, I _think_ the e2e tests will still pass even if the code doesn't work, so I wanted someone to verify manually that the `-j DROP` rule is getting hit, as opposed to the traffic getting dropped for unrelated reasons.

**Comment by danwinship** [comment]
so actually, I don't think these rules are ever relevant with GCP, and I need to test this against MetalLB...

**Comment by danwinship** [comment]
/assign @aojea
tested now...

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Review by thockin — COMMENTED** [review]
/lgtm

File: pkg/proxy/iptables/number_generated_rules_test.go [hunk 1]
```diff
@@ -248,8 +248,8 @@ func TestNumberIptablesRules(t *testing.T) {
 			},
 			services:            10,
 			epPerService:        10,
-			expectedFilterRules: 3,
-			expectedNatRules:    395,
+			expectedFilterRules: 13,
+			expectedNatRules:    385,
```
Comment (aojea) [inline pkg/proxy/iptables/number_generated_rules_test.go:252]: heh, the test worked :man_dancing:

File: pkg/proxy/iptables/proxier_test.go [hunk 2]
```diff
@@ -1651,6 +1655,7 @@ func TestTracePackets(t *testing.T) {
 		-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
 		-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
 		-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+		-A KUBE-PROXY-FIREWALL -m comment --comment "ns5/svc5:p80 traffic not accepted by KUBE-FW-NUKIZ6OKUXPJNT4C" -m tcp -p tcp -d 5.6.7.8 --dport 80 -j DROP
```
Comment (aojea) [inline pkg/proxy/iptables/proxier_test.go:1646]: what rule jumps into this rule? 
it should be referenced from INPUT, OUTPUT and FORWARD no?

Comment (aojea) [reply pkg/proxy/iptables/proxier_test.go:1646]: nevermind, you delete the extra rules for the tests in Linte 762

**Pull request merged**